presto-main/src/test/java/io/prestosql/sql/planner/TestDomainCoercer.java

```diff
@@ -0,0 +1,167 @@
+                        range(DOUBLE, 2.0, true, 3.0, true),
+                        greaterThan(DOUBLE, 4.0)),
+                true);
+        assertEquals(
+                applySaturatedCasts(original, REAL),
+                Domain.create(
+                        ValueSet.ofRanges(
+                                lessThan(REAL, (long) floatToIntBits(0.0f)),
+                                range(REAL, (long) floatToIntBits(0.0f), false, (long) floatToIntBits(1.0f), false),
+                                range(REAL, (long) floatToIntBits(2.0f), true, (long) floatToIntBits(3.0f), true),
+                                greaterThan(REAL, (long) floatToIntBits(4.0f))),
+                        true));
+    }
+
+    @Test
+    public void testOutsideTargetTypeRange()
+    {
+        assertEquals(
+                applySaturatedCasts(
+                        multipleValues(BIGINT, ImmutableList.of(1L, 10000000000L, -2000L)), SMALLINT),
+                multipleValues(SMALLINT, ImmutableList.of(1L, -2000L)));
+
+        assertEquals(
+                applySaturatedCasts(
+                        Domain.create(
+                                ValueSet.ofRanges(range(DOUBLE, 0.0, true, ((double) Float.MAX_VALUE) * 10, true)),
+                                true),
+                        REAL),
+                Domain.create(
+                        ValueSet.ofRanges((greaterThanOrEqual(REAL, (long) floatToIntBits(0.0f)))),
+                        true));
+
+        assertEquals(
+                applySaturatedCasts(
+                        Domain.create(
+                                ValueSet.ofRanges(
+                                        range(DOUBLE, ((double) Float.MAX_VALUE) * -2, true, ((double) Float.MAX_VALUE) * 10, true),
+                                        range(DOUBLE, 0.0, false, 1.0, false)),
+                                true),
+                        REAL),
+                Domain.all(REAL));
+
+        assertEquals(
+                applySaturatedCasts(
+                        Domain.create(
+                                ValueSet.ofRanges(
+                                        range(BIGINT, ((long) Integer.MAX_VALUE) * -2, false, ((long) Integer.MAX_VALUE) * 10, false),
+                                        range(BIGINT, 2L, true, 3L, true)),
+                                true),
+                        INTEGER),
+                Domain.all(INTEGER));
+    }
+
+    @Test
+    public void testTruncatedCoercedValue()
+    {
+        assertEquals(
+                applySaturatedCasts(
+                        Domain.create(
+                                ValueSet.ofRanges(
+                                        range(createDecimalType(6, 3), 123456L, true, 234567L, true)),
+                                true),
+                        createDecimalType(6, 1)),
+                Domain.create(
+                        ValueSet.ofRanges(greaterThan(createDecimalType(6, 1), 1234L)),
```

Comment:
upper bound is missing, but should be there

Reply:
fixed now